include/ensmallen_bits/eve/eve.hpp

```diff
@@ -0,0 +1,166 @@
+/**
+ * @file eve.hpp
+ * @author Marcus Edel
+ *
+ * Eve: a gradient based optimization method with Locally
+ * and globally adaptive learning rates.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_EVE_EVE_HPP
+#define ENSMALLEN_EVE_EVE_HPP
+
+namespace ens {
+
+/**
+ * Eve is a stochastic gradient based optimization method with locally and
+ * globally adaptive learning rates. Stochastic Gradient Descent is a
+ * technique for minimizing a function which can be expressed as a sum of other
+ * functions.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{Koushik2016,
+ *   author  = {Jayanth Koushik and Hiroaki Hayashi},
+ *   title   = {Improving Stochastic Gradient Descent with Feedback},
+ *   journal = {CoRR},
+ *   year    = {2016},
+ *   url     = {http://arxiv.org/abs/1611.01505}
+ * }
+ *
+ * For EVE to work, a DecomposableFunctionType template parameter is required.
+ * This class must implement the following function:
+ *
+ *   size_t NumFunctions();
+ *   double Evaluate(const arma::mat& coordinates,
+ *                   const size_t i,
+ *                   const size_t batchSize);
+ *   void Gradient(const arma::mat& coordinates,
+ *                 const size_t i,
+ *                 arma::mat& gradient,
+ *                 const size_t batchSize);
+ *
+ * NumFunctions() should return the number of functions (\f$n\f$), and in the
+ * other two functions, the parameter i refers to which individual function (or
+ * gradient) is being evaluated.  So, for the case of a data-dependent function,
+ * such as NCA, NumFunctions() should return the number of points in the
+ * dataset, and Evaluate(coordinates, 0) will evaluate the objective function on
+ * the first point in the dataset (presumably, the dataset is held internally in
+ * the DecomposableFunctionType).
+ */
+class EVE
```

Comment:
Do you think we should just call it "Eve" not "EVE"?  THe paper doesn't suggest the name as an acronym, just a reference to 'Adam and Eve' I guess.